### PR TITLE
Toolchain+CI: Decouple the LLVM toolchain from GNU Binutils

### PR DESCRIPTION
**Pull request by BertalanD**
**Toolchain: Backport --update-section support to llvm-objcopy**

This commit backports the LLVM commit that adds support for the
`--update-section` flag to llvm-objcopy. We use this feature of GNU
objcopy to embed the symbol map in the kernel.

The corresponding LLVM Phabricator Differential Revision can be found
here: https://reviews.llvm.org/D112116

This patch is identical to the upstream commit, except for two hunks
that had to be changed as they didn't apply cleanly.

**Toolchain: Don't use GNU objcopy in the LLVM toolchain**

Our build of LLVM's objcopy now supports the single missing feature
(--update-section) that previously forced us to use the one from GNU
Binutils. This means that there is no reason anymore to build Binutils
alongside LLVM's tools.

**Toolchain: Do not overwrite libraries with stubs when rebuilding LLVM**
We erroneously appended ".so" after the base name for the library,
so we ended up checking for the existence of e.g. `libc.so.so`,
which obviously didn't exist, so we overwrote the existing libraries
when we rebuilt the toolchain.

**CI: Hash only relevant files for the key of the compiler cache**

In the last few commits, a second patch was added to the LLVM toolchain,
and it no longer uses our binutils patch. This commit changes the CI
cache keys accordingly, in order to prevent unnecessary rebuilds of both
toolchains when only one is changed.

The Clang toolchain's cache now only takes into account patches that
begin with `llvm`, and the GNU toolchain excludes those from the hash
calculation. We now also hash the two CMake cache files that we use for
building LLVM and its runtime libraries.

### CODE REVIEW

File: Meta/Azure/Caches.yml [hunk 1]
```diff
@@ -18,13 +18,13 @@ steps:
     - ${{ if eq(parameters.toolchain, 'clang') }}:
       - task: Cache@2
         inputs:
-          key: '"toolchain" | "x86_64" | Toolchain/BuildClang.sh | Toolchain/Patches/*[!gcc].patch | Userland/Libraries/LibC/**/*.h | Userland/Libraries/LibPthread/**/*.h'
+          key: '"toolchain" | "x86_64" | Toolchain/BuildClang.sh | Toolchain/Patches/llvm*.patch | Toolchain/CMake/*.cmake | Userland/Libraries/LibC/**/*.h | Userland/Libraries/LibPthread/**/*.h'
```
Comment (ADKaster) [inline Meta/Azure/Caches.yml:21]: This also makes the clang Toolchain hash independent of the gdb patch. Irrelevant for CI, though it raises two questions

whether we should build gdb separately from the gcc Toolchain

whether we should exclude that patch from the gcc Toolchain cache as well

File: Toolchain/BuildClang.sh [hunk 2]
```diff
@@ -209,43 +198,13 @@ pushd "$DIR/Tarballs"
...
-            git init > /dev/null
-            git add . > /dev/null
-            git commit -am "BASE" > /dev/null
-            git apply "$DIR"/Patches/binutils.patch > /dev/null
-        else
-            patch -p1 < "$DIR/Patches/binutils.patch" > /dev/null
-        fi
-         $MD5SUM "$DIR/Patches/binutils.patch" > .patch.applied
+        $MD5SUM "$DIR/Patches/llvm.patch" >> .patch.applied
+        $MD5SUM "$DIR/Patches/llvm.patch" >> .patch.applied
```
[the remark this answers is not in the document]
Comment (BertalanD) [reply Toolchain/BuildClang.sh]: BSD/Darwin `md5` and `md5sum` can apparently hash multiple files at once, so I'll just use that.

File: Toolchain/BuildClang.sh [hunk 3]
```diff
@@ -208,44 +207,13 @@ pushd "$DIR/Tarballs"
             git init > /dev/null
             git add . > /dev/null
             git commit -am "BASE" > /dev/null
+            git am "$DIR"/Patches/llvm-backport-objcopy-update-section.patch > /dev/null
```
Comment (ADKaster) [inline Toolchain/BuildClang.sh:210]: any reason this one is git am and the llvm.patch is git apply?

Comment (BertalanD) [reply Toolchain/BuildClang.sh:210]: The two patches aren't in the same format.

The objcopy patch is in the `git format-patch` format, while `llvm.patch` is a simple diff file. `git am` can't apply diffs, while `git apply` discards the commit information.

**Pull request merged**